### PR TITLE
Distance between points given in geodetic coordinates

### LINKED ISSUE #154: Compute distances in geodetic coordinates
**Description of the desired feature**

Currently Harmonica has functions to compute distances between points defined in Cartesian coordinates or in spherical geocentric coordinates. Would be nice if we add a function to compute Euclidean distances between points whose coordinates are defined in geodetic (ellipsoidal) coordinates (don't confuse it with great circle distances).
Worth noting that the geodetic coordinates depend on the reference ellipsoid, so we would need to pass a `boule.Ellipsoid` to the new function along with the coordinates of the two points.
This may allow to compute gravitational and magnetic effects of point masses directly on geodetic coordinates without the need to convert them into spherical. Also it could enable us to build a separate equivalent layer class, a `EQLHarmonicGeodetic` for example, that interpolates harmonic data given in geodetic coordinates.

Some articles that may give insights of how to compute such distances are:
- Roussel et al. (2015). Complete gravity field of an ellipsoidal prism by Gauss–Legendre quadrature. doi: https://doi.org/10.1093/gji/ggv438
- Vajada et al. (2008). Global ellipsoid-referenced topographic, bathymetric and stripping corrections to gravity disturbance. doi: https://doi.org/10.1007/s11200-008-0003-5

**Are you willing to help implement and maintain this feature?** Yes

### PR DESCRIPTION
**Pull request by santisoler**
Add function to compute the Euclidean distance between two points given in geodetic coordinates.
The new function makes use of the closed-form formula given by [Vajda (2004)](https://d1wqtxts1xzle7.cloudfront.net/45194692/On_evaluation_of_Newton_integrals_in_geo20160429-31556-16b23iv.pdf?1461926098=&response-content-disposition=inline%3B+filename%3DOn_evaluation_of_Newton_integrals_in_geo.pdf&Expires=1594246043&Signature=G0FRusq9DVc1eSllnrFGZUufrskJdDQgXpOn8pzITqSQVvnwflK2a9nSZqUbfqD2msqsDEVbUiaNeUKEeXYentp3GiMmEXE~P-Dt9KddO0-AcaXf-WjTUWeJb5wmnFKFAFANWhNS4x9IlVY2NdqPP2U-Ide6KxA7yNGN-JhQ1i-OvQEG8dSAgvYJTDKhdodlWcLmn1zYtlDGtde3IMEWT2JpB-JzoXRF8aJbEMTKXEGvnb4FMgGxWsfsCmFry67Zu3pAebeRrvGhY-JYuODa2bGqr8nu8CyO0KuOV5eoVFLI2zFZWIXwDH1njr~2kLhvJwBMm9q4r7joTRAPvnmcbw__&Key-Pair-Id=APKAJLOHF5GGSLRBV4ZA).
Add test function that compares the distance between two points, with the same distance computed by converting them to a spherical coordinate system and then using the `distance_spherical` function.

Fixes #154 

**Reminders**:

- [ ] Run `make format` and `make check` to make sure the code follows the style guide.
- [ ] Add tests for new features or tests that would have caught the bug that you're fixing.
- [ ] Add new public functions/methods/classes to `doc/api/index.rst` and the base `__init__.py` file for the package.
- [ ] Write detailed docstrings for all functions/classes/methods. It often helps to design better code if you write the docstrings first.
- [ ] If adding new functionality, add an example to the docstring, gallery, and/or tutorials.
- [ ] Add your full name, affiliation, and [ORCID](https://orcid.org) (optional) to the `AUTHORS.md` file (if you haven't already) in case you'd like to be listed as an author on the [Zenodo](https://zenodo.org/communities/fatiando) archive of the next release.

### CODE REVIEW

**Comment by santisoler** [comment]
The function relies on the `boule.Ellipsoid._prime_vertical_radius()` method from https://github.com/fatiando/boule/pull/35.

File: doc/references.rst [hunk 1]
```diff
@@ -12,4 +12,5 @@ References
 .. [Hofmann-WellenhofMoritz2006] Hofmann-Wellenhof, B., & Moritz, H. (2006). Physical Geodesy (2nd, corr. ed. 2006 edition ed.). Wien ; New York: Springer.
 .. [Nagy2000] Nagy, D., Papp, G. & Benedek, J.(2000). The gravitational potential and its derivatives for the prism. Journal of Geodesy 74: 552. doi:`10.1007/s001900000116 <https://doi.org/10.1007/s001900000116>`__
 .. [Nagy2002] Nagy, D., Papp, G. & Benedek, J.(2000). Corrections to "The gravitational potential and its derivatives for the prism". Journal of Geodesy (2002) 76: 475. doi:`10.1007/s00190-002-0264-7 <https://doi.org/10.1007/s00190-002-0264-7>`__
+.. [Vajda2004] Vajda, P., Vaníček, P., Novák, P. and Meurers, B. (2004). On evaluation of Newton integrals in geodetic coordinates: Exact formulation and spherical approximation. Contributions to Geophysics and Geodesy.
```
Comment (leouieda) [inline doc/references.rst:15]:
```suggestion
.. [Vajda2004] Vajda, P., Vaníček, P., Novák, P. and Meurers, B. (2004). On evaluation of Newton integrals in geodetic coordinates: Exact formulation and spherical approximation. Contributions to Geophysics and Geodesy, 34(4), 289-314.
```

File: harmonica/forward/utils.py [hunk 2]
```diff
@@ -1,37 +1,44 @@
...
     coordinate_system : str (optional)
         Coordinate system of the coordinates of the computation points and the
         point masses.
-        Available coordinates systems: ``cartesian``, ``spherical``.
-        Default ``cartesian``.
+        Available coordinates systems: ``cartesian``, ``spherical`` and
+       ``geodetic``. Default ``cartesian``.
+    ellipsoid : :class:`boule.Ellipsoid`
+        Reference ellipsoid for points coordinates. Ignored if
+        ``coordinate_system`` is not ``geodetic``. Default ``None``.
```
Comment (leouieda) [inline harmonica/forward/utils.py:41]:
```suggestion
        ``coordinate_system`` is not ``"geodetic"``. Default ``None``.
```

File: harmonica/forward/utils.py [hunk 3]
```diff
@@ -178,3 +187,129 @@ def distance_spherical_core(
...
     dist = np.sqrt((radius - radius_p) ** 2 + 2 * radius * radius_p * (1 - cospsi))
     return dist, cospsi, coslambda
+
+
+def distance_geodetic(point_p, point_q, ellipsoid):
+    """
+    Calculate the distance between two points in geodetic coordinates
+
+    Computes the Euclidean distance between two points given in geodetic
+    coordinates using the closed-form formula given by [Vajda2004]
```
Comment (leouieda) [inline harmonica/forward/utils.py:197]:
```suggestion
    coordinates using the closed-form formula given by [Vajda2004]_.
```

File: harmonica/forward/utils.py [hunk 4]
```diff
@@ -178,3 +187,129 @@ def distance_spherical_core(
...
+    term_4 = (upward_sum_p - ecc_sq * prime_vertical_radius_p) ** 2 * sinphi_p ** 2
+    term_5 = (upward_sum - ecc_sq * prime_vertical_radius) ** 2 * sinphi ** 2
+    term_6 = (
+        2
+        * (upward_sum_p - ecc_sq * prime_vertical_radius_p)
+        * (upward_sum - ecc_sq * prime_vertical_radius)
+        * sinphi
+        * sinphi_p
+    )
+    distance_sq = term_1 + term_2 - term_3 + term_4 + term_5 - term_6
```
Comment (leouieda) [inline harmonica/forward/utils.py:314]: Any particular reason to split this up like this and not just have each component in a line for the sum?

Comment (santisoler) [reply harmonica/forward/utils.py:314]: Nope... I split the computation in terms in order to order my head when transcribing the equations, but sure we might want to shrink them into a single line.

Comment (santisoler) [reply harmonica/forward/utils.py]: (effects of late evening coding)

File: harmonica/forward/utils.py [hunk 5]
```diff
@@ -1,37 +1,44 @@
...
 """
 import numpy as np
 from numba import jit
 
 
-def distance(point_p, point_q, coordinate_system="cartesian"):
+def distance(point_p, point_q, coordinate_system="cartesian", ellipsoid=None):
     """
-    Distance between two points in Cartesian or spherical coordinates
+    Distance between two points in Cartesian spherical or geodetic coordinates
```
Comment (leouieda) [inline harmonica/forward/utils.py:10]:
```suggestion
    Distance between two points in Cartesian, spherical, or geodetic coordinates
```

File: harmonica/forward/utils.py [hunk 6]
```diff
@@ -178,3 +187,130 @@ def distance_spherical_core(
...
+        Euclidean distance between ``point_p`` and ``point_q``.
+
+    Example
+    -------
+
+    >>> import boule as bl
+    >>> ellipsoid = bl.WGS84
+    >>> point_p = (-72.3, -33.3, 644)
+    >>> point_q = (-70.1, -31.6, 1024)
+    >>> distance_geodetic(point_p, point_q, ellipsoid)
```
Comment (leouieda) [inline harmonica/forward/utils.py:228]:
```suggestion
    >>> point_p = (-72.3, -33.3, 644)
    >>> point_q = (-70.1, -31.6, 1024)
    >>> distance = distance_geodetic(point_p, point_q, ellipsoid)
    >>> print("{:.2f} m".format(distance))
    279878.84 m
```

Comment (santisoler) [reply harmonica/forward/utils.py:228]: I was waiting for Boule 0.2 to be out in order to get the expected example result and to run the tests to see if the closed-form formula works fine.

**Pull request merged**